src/lib/core/components/ScatterplotRsquareTable.tsx

```diff
@@ -0,0 +1,140 @@
+import React from 'react';
+import {
+  XYPlotDataSeries,
+  XYPlotData,
+  FacetedData,
+} from '@veupathdb/components/lib/types/plots';
+import { VariableTreeNode } from '../types/study';
+
+type FacetDataProps = FacetedData<XYPlotData>['facets'];
+
+// introduce discriminated union
+type TypedScatterplotRsquareData =
+  | {
+      isFaceted: false;
+      data?: XYPlotDataSeries[];
+    }
+  | {
+      isFaceted: true;
+      // FacetDataProps is already array, so not FacetDataProps[]
+      data?: FacetDataProps;
+    };
+
+export interface Props {
+  typedData: TypedScatterplotRsquareData;
+  overlayVariable?: VariableTreeNode;
+  facetVariable?: VariableTreeNode;
+}
+
+export function ScatterplotRsquareTable({
+  typedData,
+  overlayVariable,
+  facetVariable,
+}: Props) {
+  // non-facet or facet data
+  const typedFilteredData: TypedScatterplotRsquareData = !typedData.isFaceted
+    ? {
+        isFaceted: false,
+        data:
+          overlayVariable != null
+            ? typedData.data?.filter((data) =>
+                data.name?.includes(', Best fit')
+              )
+            : typedData.data?.filter((data) => data.name?.includes('Best fit')),
+      }
+    : {
+        isFaceted: true,
+        data: typedData.data?.filter((element) => element.data != null),
+      };
+
+  if (!typedFilteredData.isFaceted) {
+    return (
+      <div className={'ScatterRsquareTable'}>
+        <table>
+          <tbody>
+            <tr>
+              <th>
+                {overlayVariable != null ? overlayVariable.displayName : 'Name'}
+              </th>
+              <th className="numeric">
+                R<sup>2</sup> (Best fit)
+              </th>
+            </tr>
+            {typedFilteredData.data?.map((data) => (
+              <tr key={data.name}>
+                <td>{data?.name?.split(', Best fit')[0]}</td>
+                <td>{data.r2 ?? 'N/A'}</td>
+              </tr>
+            ))}
+          </tbody>
+        </table>
+      </div>
+    );
+  } else {
+    return (
+      <div
+        className={'ScatterRsquareTable'}
+        style={{ maxHeight: 250, overflowX: 'hidden', overflowY: 'auto' }}
+      >
+        <table>
+          <tbody>
+            <tr>
+              <th>{facetVariable?.displayName}</th>
+              {overlayVariable != null && (
+                <th>{overlayVariable.displayName}</th>
+              )}
+              <th className="numeric">
+                R<sup>2</sup> (Best fit)
+              </th>
+            </tr>
+            {typedFilteredData.data?.map((data) => (
+              <>
```

Comment:
You'll want to give a key to each of these [React fragments](https://reactjs.org/docs/fragments.html). To do this, you'll need to switch from using the ["short" <> syntax](https://reactjs.org/docs/fragments.html) (which can't receive a `key`) to the `<React.Fragment>` syntax, which [*can* receive a `key`](https://reactjs.org/docs/fragments.html#keyed-fragments).

```suggestion
              <React.Fragment key={data.label}>
```

Reply:
In general, whenever you're `.map`ping to an array of React elements (even `<>`), those elements should be given unique keys.